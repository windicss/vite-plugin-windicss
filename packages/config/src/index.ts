import fs from 'node:fs'
import path from 'node:path'
import _debug from 'debug'
import type { JITI } from 'jiti'
import _jiti from 'jiti'
import type { FullConfig as WindiCssOptions } from 'windicss/types/interfaces'

let jiti: JITI

export type { WindiCssOptions }

export const defaultConfigureFiles = [
  'windi.config.ts',
  'windi.config.mts',
  'windi.config.js',
  'windi.config.mjs',
  'windi.config.cjs',

  'windicss.config.ts',
  'windicss.config.mts',
  'windicss.config.js',
  'windicss.config.mjs',
  'windicss.config.cjs',

  'tailwind.config.ts',
  'tailwind.config.mts',
  'tailwind.config.js',
  'tailwind.config.mjs',
  'tailwind.config.cjs',
]

export interface LoadConfigurationOptions {
  /**
   * Name for debug
   *
   * @default 'windicss-config'
   * @internal
   */
  name?: string
  /**
   * Config object or path
   */
  config?: WindiCssOptions | string
  /**
   * CWD
   *
   * @default process.cwd
   * @internal
   */
  root?: string
  /**
   * A list of filename of paths to search of config files
   */
  configFiles?: string[]
  /**
   * On loading configuration error
   *
   * @default [throw error]
   */
  onConfigurationError?: (error: unknown) => void
  /**
   * On configure file not found
   *
   * @default [emit warning]
   */
  onConfigurationNotFound?: (filepath: string) => void
}

export function loadConfiguration(options: LoadConfigurationOptions): {
  config: WindiCssOptions
  filepath?: string
  error?: unknown
} {
  if (!jiti)
    jiti = _jiti(process.cwd(), { requireCache: false, cache: false, v8cache: false })

  let resolved: WindiCssOptions = {}
  let configFilePath: string | undefined
  let error: unknown

  const {
    name = 'windicss-config',
    config,
    root = process.cwd(),
    configFiles: configureFiles = defaultConfigureFiles,
    onConfigurationError = (e) => {
      throw e
    },
    onConfigurationNotFound = (path) => {
      console.warn(`[${name}] config file "${path}" not found, ignored`)
    },
  } = options

  const debugConfig = _debug(`${name}:config`)

  if (typeof config === 'string' || !config) {
    if (!config) {
      for (const name of configureFiles) {
        const tryPath = path.resolve(root, name)
        if (fs.existsSync(tryPath)) {
          configFilePath = tryPath
          break
        }
      }
    }
    else {
      configFilePath = path.resolve(root, config)
      if (!fs.existsSync(configFilePath)) {
        onConfigurationNotFound(config)
        configFilePath = undefined
      }
    }

    if (configFilePath) {
      try {
        debugConfig('loading from ', configFilePath)

        resolved = jiti(configFilePath)
        if (resolved.default)
          resolved = resolved.default
      }
      catch (e) {
        error = e
        configFilePath = undefined
        resolved = {}
        onConfigurationError?.(e)
      }
    }
  }
  else {
    resolved = config
  }

  debugConfig(resolved)

  return {
    error,
    config: resolved,
    filepath: configFilePath,
  }
}
