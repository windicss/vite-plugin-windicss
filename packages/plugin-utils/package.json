{
  "name": "@windicss/plugin-utils",
  "version": "1.8.6",
  "description": "Common utils for building integrations of Windi CSS",
  "author": "antfu <anthonyfu117@hotmail.com>",
  "license": "MIT",
  "funding": "https://github.com/sponsors/antfu",
  "homepage": "https://github.com/antfu/vite-plugin-windicss",
  "repository": {
    "type": "git",
    "url": "https://github.com/antfu/vite-plugin-windicss"
  },
  "bugs": "https://github.com/antfu/vite-plugin-windicss/issues",
  "keywords": [
    "windicss",
    "tailwindcss"
  ],
  "exports": {
    ".": {
      "require": "./dist/index.js",
      "import": "./dist/index.mjs"
    },
    "./*": "./*"
  },
  "main": "dist/index.js",
  "module": "dist/index.mjs",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsup src/index.ts --dts --format cjs,esm --no-splitting --external sucrase/register/ts,pug,windicss/utils/style,windicss/utils/parser,windicss/utils",
    "dev": "npm run build -- --watch",
    "prepublishOnly": "npm run build"
  },
  "dependencies": {
    "@antfu/utils": "^0.5.2",
    "@windicss/config": "workspace:*",
    "debug": "^4.3.4",
    "fast-glob": "^3.2.11",
    "magic-string": "^0.26.2",
    "micromatch": "^4.0.5",
    "windicss": "^3.5.6"
  },
  "devDependencies": {
    "pug": "^3.0.2"
  }
}
