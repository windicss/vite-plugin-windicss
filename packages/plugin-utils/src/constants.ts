export const defaultAlias: Record<string, TagNames> = {
  'router-link': 'a',
}

export const preflightTags = ['html', 'body', 'div']

export const htmlTags = [
  'html',
  'body',
  'div',
  'a',
  'abbr',
  'address',
  'area',
  'article',
  'aside',
  'audio',
  'base',
  'basefont',
  'bdo',
  'blink',
  'blockquote',
  'br',
  'button',
  'canvas',
  'caption',
  'center',
  'col',
  'colgroup',
  'command',
  'comment',
  'datalist',
  'dd',
  'del',
  'details',
  'dir',
  'dl',
  'dt',
  'embed',
  'fieldset',
  'figure',
  'b',
  'big',
  'i',
  'small',
  'tt',
  'font',
  'footer',
  'form',
  'frame',
  'frameset',
  'head',
  'header',
  'hgroup',
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6',
  'hr',
  'isindex',
  'iframe',
  'ilayer',
  'img',
  'input',
  'ins',
  'keygen',
  'keygen',
  'label',
  'layer',
  'legend',
  'li',
  'link',
  'map',
  'mark',
  'marquee',
  'menu',
  'meta',
  'meter',
  'multicol',
  'nav',
  'nobr',
  'noembed',
  'noframes',
  'noscript',
  'object',
  'ol',
  'optgroup',
  'option',
  'output',
  'p',
  'param',
  'cite',
  'code',
  'dfn',
  'em',
  'kbd',
  'samp',
  'strong',
  'var',
  'plaintext',
  'pre',
  'progress',
  'q',
  'ruby',
  'script',
  'section',
  'select',
  'spacer',
  'span',
  's',
  'strike',
  'style',
  'sub',
  'sup',
  'svg',
  'table',
  'tbody',
  'td',
  'textarea',
  'tfoot',
  'th',
  'thead',
  'time',
  'title',
  'tr',
  'u',
  'ul',
  'video',
  'wbr',
  'wbr',
  'xmp',
] as const

export type TagNames = (typeof htmlTags)[number]
