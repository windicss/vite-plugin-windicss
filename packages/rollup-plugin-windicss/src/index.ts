import { Plugin } from 'rollup'
import _debug from 'debug'
import { createUtils, UserOptions } from '@windicss/plugin-utils'
import { createVirtualModuleLoader } from '../../shared/virtual-module'

const NAME = 'rollup-plugin-windicss'

const debug = {
  hmr: _debug(`${NAME}:hmr`),
  css: _debug(`${NAME}:transform:css`),
  group: _debug(`${NAME}:transform:group`),
  alias: _debug(`${NAME}:transform:alias`),
}

function WindiCssRollupPlugin(userOptions: UserOptions = {}): Plugin[] {
  const utils = createUtils(
    userOptions,
    {
      name: NAME,
      onConfigurationError(e) {
        throw e
      },
    },
  )
  const plugins: Plugin[] = []

  plugins.push({
    name: `${NAME}:alias`,
    async transform(code, id) {
      await utils.ensureInit()
      if (!utils.isDetectTarget(id))
        return
      debug.alias(id)
      return utils.transformAlias(code, true)
    },
  })

  if (userOptions.transformGroups !== false) {
    plugins.push({
      name: `${NAME}:groups`,
      async transform(code, id) {
        await utils.ensureInit()
        if (!utils.isDetectTarget(id))
          return
        debug.group(id)
        return utils.transformGroups(code, true)
      },
    })
  }

  plugins.push({
    name: `${NAME}:entry`,

    ...createVirtualModuleLoader({ utils }),
  })

  if (userOptions.transformCSS !== false) {
    plugins.push({
      name: `${NAME}:css`,
      async transform(code, id) {
        await utils.ensureInit()
        if (!utils.isCssTransformTarget(id))
          return
        debug.css(id)
        return utils.transformCSS(code, id)
      },
    })
  }

  return plugins
}

export * from '@windicss/plugin-utils'
export default WindiCssRollupPlugin
