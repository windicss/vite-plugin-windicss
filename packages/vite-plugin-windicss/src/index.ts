import { resolve } from 'node:path'
import type { Plugin, ResolvedConfig, ViteDevServer } from 'vite'
import _debug, { log } from 'debug'
import type { UserOptions, WindiPluginUtils, WindiPluginUtilsOptions } from '@windicss/plugin-utils'
import { createUtils } from '@windicss/plugin-utils'
import { MODULE_ID_VIRTUAL_PREFIX, createVirtualModuleLoader } from '../../shared/virtual-module'
import { createDevtoolsPlugin } from './devtools'
import { NAME } from './constants'
import { getCssModules, reloadChangedCssModules } from './modules'

const debug = {
  hmr: _debug(`${NAME}:hmr`),
  css: _debug(`${NAME}:transform:css`),
  group: _debug(`${NAME}:transform:group`),
  alias: _debug(`${NAME}:transform:alias`),
  memory: _debug(`${NAME}:memory`),
}

function VitePluginWindicss(userOptions: UserOptions = {}, utilsOptions: WindiPluginUtilsOptions = {}): Plugin[] {
  let utils: WindiPluginUtils
  // let viteConfig: ResolvedConfig
  let server: ViteDevServer | undefined
  let viteConfig: ResolvedConfig

  const plugins: Plugin[] = []

  // transform alias
  plugins.push({
    name: `${NAME}:alias`,
    enforce: 'pre',
    configResolved(_config) {
      viteConfig = _config
    },
    async transform(code, id) {
      await utils.ensureInit()
      if (!utils.isDetectTarget(id))
        return
      debug.alias(id)
      return utils.transformAlias(code, !!viteConfig.build.sourcemap)
    },
  })

  // Utilities grouping transform
  if (userOptions.transformGroups !== false) {
    plugins.push({
      name: `${NAME}:groups`,
      enforce: 'pre',
      async transform(code, id) {
        await utils.ensureInit()
        if (!utils.isDetectTarget(id))
          return
        debug.group(id)
        return utils.transformGroups(code, !!viteConfig.build.sourcemap)
      },
    })
  }

  // exposing api
  plugins.push({
    name: NAME,
    get api() {
      return utils
    },
  })

  // CSS Entry via virtual module
  plugins.push({
    name: `${NAME}:entry`,
    enforce: 'post',

    configureServer(_server) {
      server = _server
    },

    async configResolved(_config) {
      // viteConfig = _config
      utils = utilsOptions.utils ?? createUtils(userOptions, {
        name: NAME,
        root: _config.root,
        onConfigurationError(e) {
          if (_config.command === 'build') {
            throw e
          }
          else {
            console.error(`[${NAME}] Error on loading configurations`)
            console.error(e)
          }
        },
        ...utilsOptions,
      })
      await utils.ensureInit()
    },

    ...createVirtualModuleLoader({
      get utils() { return utils },
      get inHmr() {
        return Boolean(server)
      },
    }) as Omit<Plugin, 'name'>,
  })

  let _cssReloadTask: any
  function reloadCssModules(server: ViteDevServer) {
    clearTimeout(_cssReloadTask)
    _cssReloadTask = setTimeout(() => {
      reloadChangedCssModules(server, utils)
    }, 1)
  }

  // HMR
  plugins.push({
    name: `${NAME}:hmr`,
    apply: 'serve',
    enforce: 'pre',

    async configureServer(_server) {
      server = _server

      await utils.ensureInit()
      if (utils.configFilePath)
        server.watcher.add(utils.configFilePath)

      // NOTE: Track changes to the files so that they are re-scanned as needed.
      // Added files are only detected if the user explicitly enables globbing.
      const supportsGlobs = server.config.server.watch?.disableGlobbing === false
      server.watcher.add(supportsGlobs ? utils.globs : await utils.getFiles())
    },

    async handleHotUpdate({ server, file, read }) {
      // resolve normalized file path to system path
      if (resolve(file) === utils.configFilePath) {
        debug.hmr(`config file changed: ${file}`)
        await utils.init()
        setTimeout(() => {
          log('configure file changed, reloading')
          server.ws.send({ type: 'full-reload' })
        }, 0)
        return getCssModules(server)
      }

      if (!utils.isDetectTarget(file))
        return

      utils.extractFile(await read(), file, true)
        .then((changed) => {
          if (changed) {
            debug.hmr(`refreshed by ${file}`)
            reloadCssModules(server)
          }
        })
    },
  })

  const { transformCSS: transformCSSOptions = true } = userOptions

  const transformCSS = (code: string, id: string) => utils.transformCSS(code, id, {
    onLayerUpdated() {
      if (server)
        reloadCssModules(server)
    },
  })

  // CSS transform
  if (transformCSSOptions === true) {
    plugins.push({
      name: `${NAME}:css`,
      async transform(code, id) {
        await utils.ensureInit()
        if (!utils.isCssTransformTarget(id) || id.startsWith(MODULE_ID_VIRTUAL_PREFIX))
          return
        debug.css(id)
        code = transformCSS(code, id)
        if (viteConfig.build.sourcemap) {
          return {
            code: transformCSS(code, id),
            map: { mappings: '' },
          }
        }
        else {
          return code
        }
      },
    })
  }
  else if (typeof transformCSSOptions === 'string') {
    plugins.push({
      name: `${NAME}:css`,
      enforce: transformCSSOptions,
      transform(code, id) {
        if (!utils.isCssTransformTarget(id) || id.startsWith(MODULE_ID_VIRTUAL_PREFIX))
          return
        debug.css(id, transformCSSOptions)
        code = transformCSS(code, id)
        if (viteConfig.build.sourcemap) {
          return {
            code: transformCSS(code, id),
            map: { mappings: '' },
          }
        }
        else {
          return code
        }
      },
    })
  }

  plugins.push({
    name: `${NAME}:css:svelte`,
    api: {
      sveltePreprocess: {
        style({ content, id }: { content: string; id: string }) {
          return {
            code: transformCSS(content, id),
          }
        },
      },
    },
  })

  plugins.push(...createDevtoolsPlugin({
    get utils() { return utils },
  }))

  return plugins
}

export * from '@windicss/plugin-utils'
export default VitePluginWindicss
