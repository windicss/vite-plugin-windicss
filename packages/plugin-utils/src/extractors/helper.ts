import { extname } from 'path'
import { uniq } from '@antfu/utils'
import type { Extractor, ExtractorResultDetailed } from 'windicss/types/interfaces'
import { DefaultExtractor } from './default'
import { PugExtractor } from './pug'
import { SvelteExtractor } from './svelte'

export function getDefaultExtractors() {
  const extractors: Extractor[] = [
    {
      extractor: SvelteExtractor,
      extensions: ['svelte'],
    },
  ]

  // auto detect pug
  try {
    require('pug')
    extractors.push({
      extractor: PugExtractor,
      extensions: ['vue', 'pug'],
    })
  }
  catch (e) {}

  return extractors
}

export async function applyExtractors(code: string, id?: string, extractors: Extractor[] = [], defaultExtract = DefaultExtractor) {
  let ext = id ? extname(id) : '*'
  if (ext[0] === '.')
    ext = ext.slice(1)

  const matchingExtractors = extractors
    .filter(extractor => extractor.extensions.includes(ext))
    .map(extractor => extractor.extractor)
  return Promise.all((matchingExtractors.length ? matchingExtractors : [defaultExtract])
    .map(extractor => extractor(code, id)))
    .then((results) => {
      const attributesNames = results.flatMap(v => v.attributes?.names)
      const attributesValues = results.flatMap(v => v.attributes?.values)
      return {
        tags: uniq(results.flatMap(v => v.tags)),
        ids: uniq(results.flatMap(v => v.ids)),
        classes: uniq(results.flatMap(v => v.classes)),
        attributes: (attributesNames.length || attributesValues.length)
          ? {
            names: attributesNames,
            values: attributesValues,
          }
          : undefined,
      } as ExtractorResultDetailed
    })
}
