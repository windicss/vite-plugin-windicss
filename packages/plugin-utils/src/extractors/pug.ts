import type { ExtractorResultDetailed } from 'windicss/types/interfaces'
import { DefaultExtractor } from '../extractors/default'

const regexTemplate = /<template.*?lang=['"]pug['"][^>]*?>\n([\s\S]*?\n)<\/template>/gm

export function PugExtractor(code: string, id?: string): ExtractorResultDetailed {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const Pug = require.resolve('pug') as typeof import('pug')

  const compile = (code: string) => {
    try {
      return Pug.compile(code, { filename: id })()
      // other build processes will catch pug errors
    }
    catch {}
  }

  let compiled: string | undefined

  if (id && id.match(/\.vue$/)) {
    const matches = Array.from(code.matchAll(regexTemplate))
    let tail = ''
    for (const match of matches) {
      if (match && match[1])
        tail += `\n\n${compile(match[1])}`
    }
    if (tail)
      compiled = `${code}\n\n${tail}`
  }
  else {
    compiled = compile(code)
  }

  return DefaultExtractor(compiled || code)
}
