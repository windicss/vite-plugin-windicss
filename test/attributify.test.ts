import { describe, expect, it } from 'vitest'
import { DefaultExtractor } from '../packages/plugin-utils/src/extractors/default'
import { createUtils } from '../packages/plugin-utils/src'

describe('attributify', () => {
  it('basic', () => {
    expect(DefaultExtractor(`
      <div
        p="x-4 y-2 md:x-2 lg:hover:x-8"
        border="~"
        sm="bg-red-500 bg-opacity-50 hover:bg-green-300"
        sm:hover="text-red-500 text-lg focus:bg-white"
dark:border="~ red-400"
        @click="a()"
        class="text-red-400"
      />
    `),
    ).toMatchSnapshot()
  })

  it('jsx', async () => {
    const utils = createUtils({
      config: {
        attributify: true,
      },
      safelist: ['bg-gradient-to-r'],
      preflight: false,
      scan: false,
    })
    const code = `
    <nav
      pos="fixed top-0"
      w="full"
      z="30"
      bg={props.scrolled() || props.opened() ? 'white' : 'transparent'}
      shadow={props.scrolled() ? 'md' : 'none'}
    >
      {props.children}
    </nav>
    `
    expect(DefaultExtractor(code)).toMatchSnapshot()
    await utils.ensureInit()
    await utils.extractFile(code)

    expect(await utils.generateCSS()).toMatchSnapshot()
  })
})
