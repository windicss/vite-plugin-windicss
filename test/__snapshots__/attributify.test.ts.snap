// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`attributify basic 1`] = `
Object {
  "attributes": Object {
    "names": Array [
      "p",
      "border",
      "sm",
      "sm:hover",
      "dark:border",
    ],
    "values": Array [
      "x-4 y-2 md:x-2 lg:hover:x-8",
      "~",
      "bg-red-500 bg-opacity-50 hover:bg-green-300",
      "text-red-500 text-lg focus:bg-white",
      "~ red-400",
    ],
  },
  "classes": Array [
    "<div",
    "x-4",
    "y-2",
    "md:x-2",
    "lg:hover:x-8",
    "bg-red-500",
    "bg-opacity-50",
    "hover:bg-green-300",
    "text-red-500",
    "text-lg",
    "focus:bg-white",
    "red-400",
    "a()",
    "text-red-400",
  ],
  "tags": Array [
    "div",
  ],
}
`;

exports[`attributify jsx 1`] = `
Object {
  "attributes": Object {
    "names": Array [
      "pos",
      "w",
      "z",
      "bg",
      "shadow",
    ],
    "values": Array [
      "fixed top-0",
      "full",
      "30",
      "props.scrolled() || props.opened() ? 'white' : 'transparent'",
      "props.scrolled() ? 'md' : 'none'",
    ],
  },
  "classes": Array [
    "<nav",
    "fixed",
    "top-0",
    "full",
    "props.scrolled()",
    "props.opened()",
    "white",
    "transparent",
    "props.scrolled()",
    "none",
    "props.children",
  ],
  "tags": Array [
    "nav",
  ],
}
`;

exports[`attributify jsx 2`] = `
"/* windicss layer base */

/* windicss layer components */

/* windicss layer utilities */
[bg~=\\"white\\"] {
  --tw-bg-opacity: 1;
  background-color: rgba(255, 255, 255, var(--tw-bg-opacity));
}
[bg~=\\"transparent\\"] {
  background-color: transparent;
}
.bg-gradient-to-r {
  background-image: -o-linear-gradient(left, var(--tw-gradient-stops));
  background-image: -webkit-gradient(linear, left top, right top, from(var(--tw-gradient-stops)));
  background-image: linear-gradient(to right, var(--tw-gradient-stops));
}
.fixed {
  position: fixed;
}
[pos~=\\"fixed\\"] {
  position: fixed;
}
.top-0 {
  top: 0px;
}
[pos~=\\"top-0\\"] {
  top: 0px;
}
[shadow~=\\"props.scrolled()\\"] {
  
  --tw-shadow: var(--tw-shadow-colored);
}
[shadow~=\\"?\\"] {
  
  --tw-shadow: var(--tw-shadow-colored);
}
[shadow~=\\":\\"] {
  
  --tw-shadow: var(--tw-shadow-colored);
}
[shadow~=\\"md\\"] {
  --tw-shadow: 0 4px 6px -1px rgb(0 0 0/0.1),0 2px 4px -2px rgb(0 0 0/0.1);
  --tw-shadow-colored: 0 4px 6px -1px var(--tw-shadow-color),0 2px 4px -2px var(--tw-shadow-color);
  -webkit-box-shadow: var(--tw-ring-offset-shadow,0 0 #0000),var(--tw-ring-shadow,0 0 #0000),var(--tw-shadow);
  box-shadow: var(--tw-ring-offset-shadow,0 0 #0000),var(--tw-ring-shadow,0 0 #0000),var(--tw-shadow);
}
[shadow~=\\"none\\"] {
  --tw-shadow: 0 0 #0000;
  --tw-shadow-colored: 0 0 #0000;
  -webkit-box-shadow: var(--tw-ring-offset-shadow,0 0 #0000),var(--tw-ring-shadow,0 0 #0000),var(--tw-shadow);
  box-shadow: var(--tw-ring-offset-shadow,0 0 #0000),var(--tw-ring-shadow,0 0 #0000),var(--tw-shadow);
}
[w~=\\"full\\"] {
  width: 100%;
}
[z~=\\"30\\"] {
  z-index: 30;
}"
`;
